src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,57 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    double diffOfTimeMSec = prevTime->msecsTo(myTime);
+    double diffOfTime = diffOfTimeMSec/3600000;
+    double avgCurrent = (myCurrent+prevCurrent)/2;
+
+    double changeInAmpHour = avgCurrent * diffOfTime;
+    ampHourTotal += changeInAmpHour;
+    //if(ampHourTotal > BATTERY_AMP_HOUR_CAPACITY){ampHourTotal = BATTERY_AMP_HOUR_CAPACITY;}
+    ampHourUsed = BATTERY_AMP_HOUR_CAPACITY- ampHourTotal;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(myCurrent>=0)
+    {
+        return false;
+    }
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double ampHoursLeft;
+    if(isCharging())
+    {
+        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal;
+    }
+    else
+    {
+        ampHoursLeft = ampHourTotal;
+    }
+    int current = myCurrent;
+    if(myCurrent < 0)
```

Comment:
You've already checked this condition previously with `isCharging()`, you should put the logic in these statements into where you've already checked the condition. So:
```c++
    double ampHoursLeft;
    int current = myCurrent;
    if(isCharging())
    {
        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal;
    }
    else
    {
        ampHoursLeft = ampHourTotal;
        current = -myCurrent;
    }
```